internal/workflow.go

```diff
@@ -297,6 +297,27 @@ type (
 		fn       interface{}
 		isMethod bool
 	}
+
+	UpdateDurabilityPreference int
+
+	// UpdateOptions consists of options for executing a named workflow update.
+	UpdateOptions struct {
+		// Validator is an optional (i.e. can be left nil) func with exactly the
+		// same type signature as the required update handler func but returning
+		// only a single value of type error. The implementation of this
+		// function MUST be safe and MUST NOT alter workflow state in any way.
+		// If the returned error can be cast to `interface{
+		// PreferBypassDurability() bool }` then the preference indicated by
+		// invoking PreferBypassDurability() will be passed upstream.  A panic
+		// from this function will be treated as equivalent to returning an
+		// error with a positive durability bypass preference.
+		Validator interface{}
```

Comment:
I was thinking about this a bit. Question - do we think non-durable errors will be a fairly popular thing or rare?

If we do think they would be popular, I wonder if we want to do another approach to make it more explicit. There are basically two options here: 1) Return two errors and have first be non-durable (not good though, easy to get switched up), or 2) return a custom `ValidationResolution` type which would be an interface w/ provide three helpers/impls for. With the latter, you can have `func ValidationSuccess() ValidationResolution`, `func ValidationDurableError(error) ValidationResolution`, and `func ValidationNonDurableError(error) ValidationResolution`.

> If the returned error can be cast to `interface{ PreferBypassDurability() bool }` then the preference indicated by invoking PreferBypassDurability() will be passed upstream. 
While we should keep this doc, we should provide a helper for this, e.g. `func WrapAsNonDurableUpdateError(err) error` or something.

Reply:
I honestly don't have a clue how popular it will be so not much help there, sorry.

Both approaches fulfill the abstract requirement of the feature so I'd be happy with either and I prefer to leave these design details to the sdk author. As someone who has written a fair amount of Go I have my own intuition on which is preferrable, but It think it's better as your choice.

If we do stick with just error, I agree on a helper func. I would export the bypassDurability func from errors.go

Reply:
Current direction: return error from validator, validation error or panic will always be treated as prefer-bypass since that's the only reason you created a validation func in the first place. Otherwise it would just be inline checking in the execution func.